enhancements/hypershift/disaster-recovery/hcp-etcd-backup-api.md

```diff
@@ -0,0 +1,641 @@
+    - If the PVC it's bound to a PV and contains the label/annotation for "hash validation", the controller will perform a process to validate the existant backup.
+    - TBD: valorate to include a retryPolicy
+- The controller initiates the backup process by generating the ETCD backup object in the PVC specified in the manifest.
+  - The backup process is directed to a single ETCD member to prevent conflicts during restoration.
+  - TBD: valorate to include a retryPolicy
+  - The backup object checksum should match with the ETCD cluster health checksum. If the checksums does not matches the backup will be requeued (depending on the retry policy) and marked as failed.
+  - The label/annotation for "new backup" is removed and a new label is added for "hash validation" with the proper checksum.
+- Upon completion of the backup, the controller performs post-backup tasks:
+  - Validates the backup.
+  - Sets the status conditions based on the backup process outcome.
+- The controller concludes its tasks.
+
+
+```mermaid
+graph TB
+
+subgraph Pre-Validations
+  mainValidations(Validations) --> validationBackupAlreadyDone{Backup<br>Already<br>Done?}
+  validationBackupAlreadyDone -->y0(Yes) -->f0(Finish Success)
+  validationBackupAlreadyDone -->n0(No) -->c0(Continue)
+  mainValidations --> PV
+
+  PV --> validationExists{PVC<br>Exists?}
+  validationExists --> n1("No") --> CreatePVC --> LabelForNewBackup --> c1(Continue)
+  validationExists --> y1("Yes") --> backupExists{Label?}
+
+  backupExists --> NewBackup --> c1
+  backupExists --> ValidateBackup{Checksum<br>matches?}
+  ValidateBackup --> y2("Yes") --> f2(Finish Success)
+
+  PV --> validationHasEnoughFreeSpace{Has Enough<br>FreeSpace?}
+  validationHasEnoughFreeSpace --> y4("Yes") --> c4("continue")
+  validationHasEnoughFreeSpace --> n4("No") --> f4(Finish Error)
+
+  PV --> validationHasWritePermissions{Has Write<br>Permissions?}
+  validationHasWritePermissions --> y5("Yes") --> c5("continue")
+  validationHasWritePermissions --> n5("No") --> f5(Finish Error)
+
+
+  f4 --> retryPolicy
+  f5 --> retryPolicy
+  retryPolicy --Go back--> Validations
+end
+```
+
+```mermaid
+graph TB
+subgraph ETCD Backup Process
+  mainBackupProcess(BackupProcess) --> sched{IsScheduled?}
+  sched{IsScheduled?} --> n5("No") --> rb1(RunBackup)
+  sched{IsScheduled?} --> y5("Yes") --> BackupRun{Should I<br>RunBackup?}
+  BackupRun --> y6("Yes") --> rb1
+  BackupRun --> n6("No") --> r5{Requeue}
+  r5{Requeue} --Go back--> BackupProcess
+
+  RunBackup -->  RunBackupCommand
+  RunBackupCommand --> validateBackup{BackupOk?}
+  s5 --> label1(Remove NewBackup<br>Label/Annotation on PVC)
+  s5 --> label2(Add ValidationBackup<br>Label/Annotation on PVC)
+  validateBackup --> y7("Yes") --> s5(Finish Success) --> SetStatus
+  validateBackup --> n7("No") --> f5(Finish Error) --> SetStatus
+  f5 --> retryPolicy --Go Back--> Validations
+end
+```
+
+#### ETCDUploadController
```

Comment:
Could this be a generic component that watches `config.openshift.io/v1alpha1:Backup` resources - if we extend their spec to do uploads - and works for OCP core?

Reply:
I prefer to keep both controllers separated in functionality, maybe we can signal something among them to trigger different events.